crates/transaction-pool/src/validate/eth.rs

```diff
@@ -350,8 +370,53 @@ where
             )
         }
 
-        // Checks for max cost
+        #[cfg(not(feature = "optimism"))]
         let cost = transaction.cost();
+
+        #[cfg(feature = "optimism")]
+        let cost = {
+            let block = match self
+                .client
+                .block_by_number_or_tag(reth_primitives::BlockNumberOrTag::Latest)
+            {
+                Ok(Some(block)) => block,
+                Ok(None) => {
+                    return TransactionValidationOutcome::Error(
+                        *transaction.hash(),
+                        "Latest block should be found".into(),
+                    )
+                }
+                Err(err) => {
+                    return TransactionValidationOutcome::Error(*transaction.hash(), Box::new(err))
+                }
+            };
+
+            let mut encoded = reth_primitives::bytes::BytesMut::default();
+            transaction.to_recovered_transaction().encode_enveloped(&mut encoded);
```

Comment:
That's interesting - I think batching and pipeline these stages of transaction validation would be an optimization since the current interface is a single expensive operation packed into one function.

If transactions are validated in batches, these stages could be pipelined to schedule the work between e.g. querying the block, recovering, and re-encoding.